### PR TITLE
Merge events and batch update

### PR DESCRIPTION
**Pull request by peterlimg**
## Changes

- Implemented common way to merge events, and merged most of the events that could slow down the block finalization.
- Refactored some of the events table structure to avoid unnecessary data encoding/decoding. 
- Optimized to avoid doing heavy whole table updating for just updating one or two columns of data, which could slow down the db process speed severely. 
- Changed the stake pool reward distribution code in `payFee` transaction to rewards randomly selected N nodes/pools only. So as long as the randomness is distributed evenly in long run, the rewards would go to the pools evenly then. We do this because the delegate pool rewards updating event is pretty slow even after merging the events. It is the number of delegate pools slowed down the sql. Suppose each node has a max of 40 delegate pools, in the `load` network, we have 3 miners and 2 sharders, then we could have max of `5 * 40 = 200` delegate pools, which means we will need to update 200 rewards stake pool in every round. And that's just a small network with 5 nodes in total, it could be worse in large network such as the testnet. Anyway, with randomly rewards, we would only update limited N pools in each round, so the performance is basically consistent.   

## Tests
Tasks to complete before merging PR:
- [ ]  Ensure system tests are passing. If not [Run them manually](https://github.com/0chain/0chain/actions/workflows/system_tests.yml) to check for any regressions :clipboard:
- [ ]  Do any new system tests need added to test this change? do any existing system tests need updated? If so create a PR at [0chain/system_test](https://github.com/0chain/system_test)
- [ ]  Merge your system tests PR to master AFTER merging this PR

## Associated PRs (Link as appropriate):
- blobber:
- gosdk:
- system_test:
- zboxcli:
- zwalletcli:
- Other: ...

### CODE REVIEW

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | master  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | batch-update  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

**Comment by service-0chain** [comment]
Manual system tests [failure] with the following config
| config  | value |
| ------------- | ------------- |
| tests skipped  | FALSE  |
| system-tests  | batch-update  |
| gosdk  | NOT OVERRIDDEN  |
| zboxcli  | staging  |
| zwalletcli  | staging  |
| 0chain  | batch-update  |
| blobber  | staging  |
| authorizer  | staging  |
| 0box  | staging  |
| 0block  | staging  |
| 0dns  | staging  |
| explorer  | master  |
| 0proxy  | staging  |
| 0search  | staging  |

File: code/go/0chain.net/smartcontract/dbs/event/allocation.go [hunk 1]
```diff
@@ -116,13 +100,210 @@ func (edb *EventDb) GetActiveAllocsBlobberCount() (int64, error) {
...
+
+func withAllocBlobberTermsMerged() eventMergeMiddleware {
+	return withEventMerge(func(a, b *[]AllocationBlobberTerm) (*[]AllocationBlobberTerm, error) {
+		var (
+			aMap = make(map[string]AllocationBlobberTerm, len(*a))
+			pa   = *a
+			pb   = *b
+		)
+		for i, ai := range pa {
+			aMap[ai.BlobberID] = pa[i]
```
Comment (dabasov) [inline code/go/0chain.net/smartcontract/dbs/event/allocation.go:242]: can be aMap[ai.BlobberID] = ai

**Review by dabasov — APPROVED** [review]
lgtm

**Pull request merged**